File: note/연결리스트.js
class Node {
  constructor(data, next = null) {
    this.data = data;
    this.next = next; // 연결리스트의 tail 은 null을 가리킴
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.size = 0;
  }
  add(data) {
    const node = new Node(data);
    let current;

    if(this.head === null) {
      this.head = node;
    } else {
      current = this.head;
      while(current.next){  // tail 찾기
        current = current.next;
      }
      current.next = node;
    }
    this.size++;
  }

  insertAt(data, index) { // 중간 삽입
    if(index < 0 || index >= this.size) {
      return -1;
    }
    // if first
    if(index === 0) {
      this.head = new Node(data, this.head);
      this.size++
      return;
    }
    const node = new Node(data);
    let current = this.head;
    let = prev;
    let count =0;
    while(count<index) {
      prev = current;
      current = current.next;
      count++;
    }
    node.next = current;
    prev.next = node;
    this.size++;
  }

  // 데이터 값을 찾고 인덱스를 리턴
  read(data) {
    let current= this.head;
    let count = 0;
    while(current!=null) {
      if(current.data === data) {
        return count;
      }
      current = current.next;
      count++;
    }
    return -1;  // not found
  }

  remove(data) {
    let current = this.head;
    let prev = null;
    while(current!=null) {
      if(current.data === data) {
        if(prev === null) { // head 일 경우
          this.head = current.next;
        } else {
          prev.next = current.next;
        }
        this.size--;
        return current.data;
      }
      prev = current;
      current = current.next;
    }
    return -1; // not found
  }

  // 인덱스를 기반으로 제거
  removeAt(index) {
    if(index <0 || index>=this.size) {
      return -1;
    }
    let current = this.head;
    let prev;
    let count = 0;
    if(index === 0) { // head
      this.head = current.next;
      return current.data;
    } else {
      while(count<index) {
        count++;
        prev = current;
        current = current.next;
      }
      prev.next = current.next;
      current.next = null;
    }
    this.size--;
    return current.data;
  }

  sizeOfList() {
    return this.size;
  }

  printList() {
    let current = this.head;
    let result = '';
    while(current) {
      result += current.data + ' -> ';
      current = current.next;
    }
    return result + 'null';
  }
}

let aa = new LinkedList();
aa.add(10);
aa.add(20);
aa.add(30);
console.log(aa.printList());
aa.remove(20);
console.log(aa.printList());
aa.insertAt(40, 0);
console.log(aa.printList());
console.log(aa.read(10));
aa.removeAt(1);
console.log(aa.printList());

File: note/연결리스트.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class SinglyLinkedList:
    def __init__(self):
        self.head = None

    def is_empty(self):
        return self.head is None
    # 리시트 끝에 추가
    def append(self, data):
        new_node = Node(data)
        if self.is_empty():
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node
    # 리스트의 시작에 추가
    def prepend(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node
    # 특정 값을 가진 첫 번째 노드 제거
    def delete_with_value(self, data):
        if self.is_empty():
            return

        if self.head.data == data:
            self.head = self.head.next
            return

        current = self.head
        while current.next:
            if current.next.data == data:
                current.next = current.next.next
                return
            current = current.next

    def print_list(self):
        current = self.head
        while current:
            print(current.data, end=" -> ")
            current = current.next
        print("None")

# 예제 사용
sll = SinglyLinkedList()
sll.append(1)
sll.append(2)
sll.append(3)
sll.print_list()  # 출력: 1 -> 2 -> 3 -> None
sll.prepend(0)
sll.print_list()  # 출력: 0 -> 1 -> 2 -> 3 -> None
sll.delete_with_value(2)
sll.print_list()  # 출력: 0 -> 1 -> 3 -> None
